File: service/src/main/java/hu/bsstudio/safetycar/DefaultSafetyCarService.java
package hu.bsstudio.safetycar;

import hu.bsstudio.entity.TeamEntity;
import hu.bsstudio.repository.TeamRepository;
import hu.bsstudio.safetycar.model.SafetyCarFollowInformation;
import hu.bsstudio.safetycar.model.SafetyCarOvertakeInformation;
import hu.bsstudio.team.mapper.TeamModelEntityMapper;
import hu.bsstudio.team.model.DetailedTeam;
import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import reactor.core.publisher.Mono;

@RequiredArgsConstructor
public class DefaultSafetyCarService implements SafetyCarService {

    private final TeamRepository repository;

    @Setter(AccessLevel.PACKAGE)
    private TeamModelEntityMapper mapper = new TeamModelEntityMapper();

    @Override
    public Mono<DetailedTeam> safetyCarWasFollowed(final SafetyCarFollowInformation followInformation) {
        return Mono.just(followInformation)
            .map(SafetyCarFollowInformation::getTeamId)
            .flatMap(repository::findById)
            .map(entity -> updateFollowInformation(entity, followInformation))
            .flatMap(repository::save)
            .map(mapper::toModel);
    }

    @Override
    public Mono<DetailedTeam> safetyCarWasOvertaken(final SafetyCarOvertakeInformation safetyCarOvertakeInformation) {
        return Mono.just(safetyCarOvertakeInformation)
            .map(SafetyCarOvertakeInformation::getTeamId)
            .flatMap(repository::findById)
            .map(entity -> updateOvertakeInformation(entity, safetyCarOvertakeInformation))
            .flatMap(repository::save)
            .map(mapper::toModel);
    }

    private TeamEntity updateFollowInformation(final TeamEntity entity, final SafetyCarFollowInformation followInformation) {
        entity.setSafetyCarWasFollowed(followInformation.isSafetyCarFollowed());
        return entity;

    }

    private TeamEntity updateOvertakeInformation(final TeamEntity entity, final SafetyCarOvertakeInformation overtakeInformation) {
        entity.setNumberOfOvertakes(overtakeInformation.getNumberOfOvertakes());
        return entity;
    }
}

File: server/service/src/main/kotlin/hu/bsstudio/robonaut/safetycar/DefaultSafetyCarService.kt
package hu.bsstudio.robonaut.safetycar

import hu.bsstudio.robonaut.entity.TeamEntity
import hu.bsstudio.robonaut.repository.TeamRepository
import hu.bsstudio.robonaut.safetycar.model.SafetyCarFollowInformation
import hu.bsstudio.robonaut.safetycar.model.SafetyCarOvertakeInformation
import hu.bsstudio.robonaut.team.mapper.TeamModelEntityMapper
import hu.bsstudio.robonaut.team.model.DetailedTeam
import reactor.core.publisher.Mono

class DefaultSafetyCarService(
    private val repository: TeamRepository,
    private val mapper: TeamModelEntityMapper = TeamModelEntityMapper(),
) : SafetyCarService {
    override fun safetyCarWasFollowed(safetyCarFollowInformation: SafetyCarFollowInformation): Mono<DetailedTeam> {
        return Mono.just(safetyCarFollowInformation)
            .map(SafetyCarFollowInformation::teamId)
            .flatMap(repository::findById)
            .map { updateFollowInformation(it, safetyCarFollowInformation) }
            .flatMap(repository::save)
            .map(mapper::toModel)
    }

    override fun safetyCarWasOvertaken(safetyCarOvertakeInformation: SafetyCarOvertakeInformation): Mono<DetailedTeam> {
        return Mono.just(safetyCarOvertakeInformation)
            .map(SafetyCarOvertakeInformation::teamId)
            .flatMap(repository::findById)
            .map { updateOvertakeInformation(it, safetyCarOvertakeInformation) }
            .flatMap(repository::save)
            .map(mapper::toModel)
    }

    private fun updateFollowInformation(
        entity: TeamEntity,
        followInformation: SafetyCarFollowInformation,
    ): TeamEntity {
        entity.safetyCarWasFollowed = followInformation.safetyCarFollowed
        return entity
    }

    private fun updateOvertakeInformation(
        entity: TeamEntity,
        overtakeInformation: SafetyCarOvertakeInformation,
    ): TeamEntity {
        entity.numberOfOvertakes = overtakeInformation.numberOfOvertakes
        return entity
    }
}
